core/src/avm1/object.rs

```diff
@@ -145,6 +146,25 @@ pub trait TObject<'gc>: 'gc + Collect + Debug + Into<Object<'gc>> + Clone + Copy
             let mut proto = Value::Object(this);
             while let Value::Object(this_proto) = proto {
                 if this_proto.has_own_virtual(activation, name) {
+                    // some properties, e.g. TextField.text, should call an associated watcher
```

Comment:
Since this has nothing to do about the prototype chain lookup, this can be moved before the loop. This way watchers doesn't need to be handled in `ScriptObject::set_local`.

Reply:
The problem with sharing it with `set_local` is getting `old_value`. Here I assume if it's a virtual property old_value is undefined, which is not quite right in general but correct for TextField.

Naively calling `get` outside `if !self.has_own_property` fails the existing tests, and I can't see a way to do it that works.

Reply:
Hmm, getting `old_value` seems to lookup the prototype chain, but ignores virtual properties:
```as
var p = { x: 3 };
p.addProperty("x", function() { trace("get x"); });

var o = { __proto__: p };
o.watch("x", function(prop, old_val, new_val) { trace("old_val: " + old_val); });
o.x = 4; // traces "old_val: 3", but not "get x"
```

We can upgrade `get_watcher` to `call_watcher`, which does more things:
1. Get `old_value` by itself.
2. Call the watcher callback by itself.

This way `ScriptObject::call_watcher` can be implemented as follows (not tested):
```rs
fn set_local(
    &self,
    activation: &mut Activation<'_, 'gc, '_>,
    name: &str,
    value: &mut Value<'gc>,
) -> Result<(), Error<'gc>> {
    let mut result = Ok(());

    let watcher = self.0
            .read()
            .watchers
            .get(name, activation.is_case_sensitive())
            .cloned();
    if let Some(watcher) = watcher {
        let old_value = self.get_old_value(name, activation);
        let this = (*self).into();
        match watcher.call(activation, name, old_value, *value, this, Some(this)) {
            Ok(v) => *value = v,
            Err(Error::ThrownValue(e)) => {
                *value = Value::Undefined;
                result = Err(Error::ThrownValue(e));
            }
            Err(_) => *value = Value::Undefined,
        };
    }

    result
}
```

The question is what is the best way to implements `ScriptObject::get_old_value`. Exposing it through `TObject`?

Reply:
Probably.

`get` as-is passes the existing tests (the watch virtual test being disabled at the moment). I think putting it on TObject with the others is for the best. For a name, how about `get_non_virtual`? `get_stored`?

Reply:
Though the conclusion is right, your example code doesn't quite work as you intended: you need to pass a setter function to `addProperty` or it doesn't function (and the watcher needs to return the value that gets set, but that's less relevant).

Trying
```
var p = { x: 3 };
p.addProperty("x", function() { trace("get x"); }, function(value) { trace("set x to " + value); });

var o = { __proto__: p };
o.watch("x", function(prop, old_val, new_val) { trace("old_val: " + old_val); return new_val; });
o.x = 4;
```

traces

```
old_val: 3
set x to 4
```